front/ui/ui-charts/src/trackOccupancyDiagram/components/layers/OccupancyZonesLayer.tsx

```diff
@@ -17,19 +54,168 @@ const OccupancyZonesLayer = ({
+
+      const filteredOccupancyZones = sortedOccupancyZones.filter(
+        (zone) => zone.trackId === track.id
+      );
+
+      let primaryArrivalTime = 0;
+      let primaryDepartureTime = 0;
+      let lastDepartureTime = primaryDepartureTime;
+      let yPosition = OCCUPANCY_ZONE_Y_START;
+      let yOffset = Y_OFFSET_INCREMENT;
+      let zoneCounter = 0;
+      let zoneIndex = 0;
+
+      while (zoneIndex < filteredOccupancyZones.length) {
+        const zone = filteredOccupancyZones[zoneIndex];
+        const { arrivalTime, departureTime } = zone;
+
+        // * if the zone is not overlapping with any previous one, draw it in the center of the track
+        // * and reset the primary values
+        // *
+        // * if the zone is overlapping with the previous one, draw it below or above the previous one
+        // * depending on the overlapping counter
+        // *
+        // * if the zone is overlapping with the previous one and the counter is higher than the max zones
+        // * draw the remaining trains box
+        // *
+        if (arrivalTime > lastDepartureTime) {
+          // reset to initial value if the zone is not overlapping
+          yPosition = OCCUPANCY_ZONE_Y_START;
+          primaryArrivalTime = arrivalTime;
+          primaryDepartureTime = departureTime;
+          lastDepartureTime = departureTime;
+          yOffset = Y_OFFSET_INCREMENT;
+          zoneCounter = 1;
+
+          instructions.push({
+            type: 'occupancyZone',
+            zone,
+            offsetY: trackY + yPosition,
+            isSelected: zone.trainId === selectedTrainId,
+          });
+
+          zoneIndex++;
+        }
+
+        // if so and it's an even index, move it to the bottom, if it's an odd index, move it to the top
+        else if (zoneCounter < MAX_ZONES) {
+          if (arrivalTime >= primaryArrivalTime) {
+            if (zoneCounter % 2 === 0) {
+              yPosition -= yOffset;
+            } else {
+              yPosition += yOffset;
+            }
+          }
+
+          // update the last departure time if the current zone is longer
+          if (departureTime >= lastDepartureTime) lastDepartureTime = departureTime;
+
+          instructions.push({
+            type: 'occupancyZone',
+            zone,
+            offsetY: trackY + yPosition,
+            isSelected: zone.trainId === selectedTrainId,
+          });
+
+          zoneCounter++;
+          yOffset += Y_OFFSET_INCREMENT;
+          zoneIndex++;
+        }
+
+        // else, if there are too much trains:
+        else {
+          const nextIndex = filteredOccupancyZones.findIndex(
+            (filteredZone, i) => i > zoneIndex && filteredZone.arrivalTime >= lastDepartureTime
+          );
+
+          const remainingTrainsNb = nextIndex - zoneIndex;
+
+          instructions.push({
+            type: 'remainingTrains',
+            amount: remainingTrainsNb,
+            time: (zone.arrivalTime + zone.departureTime) / 2,
```

Comment:
You need to compute the centered time of the displayed trains in the pack to display the remainingTrains label correctly